[DDP] Merge work and future_work in reducer

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59510 [DDP] Remove the duplicate parseHookResult in reducer
* **#58937 [DDP] Merge work and future_work in reducer**

Remove `work` attribute from Reducer class in favor of `future_work`.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation, and created a new C++ comm hook called `_AllReduceCommHookWithDivFactor` to replace allreduce and also support handling uneven input.

#Original PR Issue: https://github.com/pytorch/pytorch/issues/41266

Differential Revision: [D28677383](https://our.internmc.facebook.com/intern/diff/D28677383/)

Commit:
[bootcamp] refactoring out Reducer.work

Remove `work` attribute from Reducer class in favor of `future_work`.

Also removed `copy_grad_to_bucket` method since now it's only one-line implementation.

#Closes: https://github.com/pytorch/pytorch/issues/41266

Differential Revision: [D28677383](https://our.internmc.facebook.com/intern/diff/D28677383/)

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1364,33 +1362,25 @@ void Reducer::finalize_backward() {
+        bucket.future_work,
+        "Expected bucket.future_work not to be null. "
+        "This may indicate that communication hook was not properly installed.");
+    bucket.future_work->wait();
```
@agolynski Although the future work calls wait() here, it still fails on Gloo backend. As shown by the error message in the test failures:
```
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/common_distributed.py", line 108, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/distributed/distributed_test.py", line 3624, in test_accumulate_gradients_no_sync_grad_is_view
    self._test_accumulate_gradients_no_sync(gradient_as_bucket_view=True)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/distributed/distributed_test.py", line 3588, in _test_accumulate_gradients_no_sync
    step_model(ddp_model, ddp_input, ddp_target)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/distributed/distributed_test.py", line 3563, in step_model
    loss.backward()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/_tensor.py", line 255, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/autograd/__init__.py", line 147, in backward
    Variable._execution_engine.run_backward(
RuntimeError: Work needs to be completed before calling result(). Should call wait() before result().
```

Could you please take a look?

SciPioneer replied:
The bug in gloo backend is fixed now.

agolynski commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -321,7 +321,7 @@ void Reducer::mark_variable_ready_dense(size_t variable_index) {
       // previous iterations, no copy is needed.
       if (!grad.is_alias_of(bucket_view)) {
-        this->copy_grad_to_bucket(grad, bucket_view);
+        bucket_view.copy_(grad);
```
where result is going to be divided by divFactor_?

SciPioneer replied:
The division is done in the implementation of `allreduce_hook`. No need to have another division. Will remove `divFactor_ ` in a separate PR.

SciPioneer replied:
It will be done in the allreduce comm hook. No need to use a separate `divFactor_`

agolynski commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -22,6 +22,19 @@ inline int64_t current_time_in_nanos() {
   return torch::autograd::profiler::getTime();
 }
 
+// FIXME: Should make this a static method of C++ comm hook, and call it there.
```
maybe worth addressing here to avoid code duplication?

SciPioneer replied:
It was a shortcut for debugging purpose. Will fix it in the next PR. This PR already has quite some noises.

agolynski commented:
Thanks! Please confirm that tests pass once https://github.com/pytorch/pytorch/pull/59398 lands

Commit:
Update on "[DDP] Refactoring out Reducer.work"


Remove `work` attribute from Reducer class in favor of `future_work`.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation.

Currently it still fails on Gloo backend.

#Original PR Issue: https://github.com/pytorch/pytorch/issues/41266

Differential Revision: [D28677383](https://our.internmc.facebook.com/intern/diff/D28677383/)

[ghstack-poisoned]

Commit:
Update on "[DDP] Refactoring out Reducer.work"


Remove `work` attribute from Reducer class in favor of `future_work`.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation.

Currently it still fails on Gloo backend.

#Original PR Issue: https://github.com/pytorch/pytorch/issues/41266

Differential Revision: [D28677383](https://our.internmc.facebook.com/intern/diff/D28677383/)

[ghstack-poisoned]

Commit:
Update on "[DDP] Refactoring out Reducer.work"


Remove `work` attribute from Reducer class in favor of `future_work`.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation, and created a new C++ comm hook called `_AllReduceCommHookWithDivFactor` to replace allreduce and also support handling uneven input.

#Original PR Issue: https://github.com/pytorch/pytorch/issues/41266

Differential Revision: [D28677383](https://our.internmc.facebook.com/intern/diff/D28677383/)

[ghstack-poisoned]

Commit:
Update on "[DDP] Refactoring out Reducer.work"


Remove `work` attribute from Reducer class in favor of `future_work`.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation.

Additionally, remove `copy_grad_to_bucket` method since now it's only one-line implementation, and created a new C++ comm hook called `_AllReduceCommHookWithDivFactor` to replace allreduce and also support handling uneven input.

#Original PR Issue: https://github.com/pytorch/pytorch/issues/41266

Differential Revision: [D28677383](https://our.internmc.facebook.com/intern/diff/D28677383/)

[ghstack-poisoned]
